src/main/kotlin/io/github/inductiveautomation/kindling/questdb/QuestDbView.kt

```diff
@@ -0,0 +1,329 @@
+        }.firstOrNull()?.parent
+            ?: throw ToolOpeningException("Unable to find a valid QuestDB root (_tab_index.d) inside the zip file.")
+
+        LOGGER.debug(foundDbPath.toString())
+
+        if (foundDbPath == tempDirectory) {
+            throw ToolOpeningException(
+                "Unsupported Zip Structure: The zip file must contain a parent folder for the database files. " +
+                    "Please zip the folder that contains the 'db' directory.",
+            )
+        }
+
+        LOGGER.debug("dbRoot found at: {}", foundDbPath)
+
+        foundDbPath
+    }
+
+    // Set up the server
+    private val pgPort: Int = ServerSocket(0).use { it.localPort }.also {
+        LOGGER.debug("Acquired port {} for QuestDB instance", it)
+    }
+
+    init {
+        // Create conf file
+        val confDir = dbRootPath.parent / "conf"
+        confDir.createDirectory()
+        val confFile = confDir / "server.conf"
+
+        val httpPort = ServerSocket(0).use { it.localPort } // web console port
+        val config = """
+            pg.net.bind.to=0.0.0.0:$pgPort
+            http.net.bind.to=0.0.0.0:$httpPort
+            pg.net.connection.timeout=0
+        """.trimIndent()
+
+        confFile.writeText(config)
+    }
+
+    private val server: ServerMain = ServerMain.create(dbRootPath.parent.toString()).apply {
+        start()
+        Socket().use { it.connect(InetSocketAddress(PG_HOST, pgPort), 15000) }
+    }
+
+    // Create a connection to the server
+    private var connection: Connection = PGSimpleDataSource().run {
+        databaseName = "qdb"
+        user = PG_USER
+        password = PG_PASS
+        portNumbers = intArrayOf(pgPort)
+
+        connection
+    }
+
+    private val pageLabel: JLabel
+
+    @Suppress("SqlResolve")
+    val tables: List<Table> = connection
+        .executeQuery("""SELECT table_name FROM tables();""")
+        .toList { resultSet ->
+            resultSet.getString("table_name")
+        }.mapNotNull { tableName ->
+            try {
+                val columns = connection
+                    .executeQuery("""SHOW COLUMNS FROM "$tableName";""")
+                    .toList { rs ->
+                        Column(
+                            name = rs.getString("column"),
+                            type = rs.getString("type"),
+                            notNull = false,
+                            defaultValue = null,
+                            primaryKey = rs.getBoolean("upsertKey"),
+                            hidden = false,
+                            _parent = { sortableTree.root },
+                        )
+                    }
+                val size: Long
+                connection.executeQuery("""SELECT diskSize FROM table_storage() WHERE tableName = '$tableName'""").use { rs ->
+                    // Move the cursor to the first row. If there's no row, default to 0.
+                    size = if (rs.next()) rs.getLong("diskSize") else 0L
```

Comment:
You could collapse this to a single expression - `val size = connection.executeQuery().use { rs.getLong() else 0L}`